winsup/cygwin/include/cygwin/exit_process.h

```diff
@@ -0,0 +1,281 @@
+
+  if (is_32_bit_os == -2 || current_is_wow == -2)
+    return FALSE;
+
+  if (!IsWow64Process (process, &is_wow))
+    return FALSE;
+
+  bitness = is_32_bit_os || is_wow ? 32 : 64;
+  return TRUE;
+}
+
+/**
+ * Terminates the process corresponding to the process ID
+ *
+ * This way of terminating the processes is not gentle: the process gets
+ * no chance of cleaning up after itself (closing file handles, removing
+ * .lock files, terminating spawned processes (if any), etc).
+ */
+static int
+exit_one_process(HANDLE process, int exit_code, DWORD pid)
+{
+  LPTHREAD_START_ROUTINE address = NULL;
+  int signo = exit_code & 0x7f;
+  BOOL is_wow;
+  int bitness;
+  switch (signo)
+  {
+  case SIGINT:
+  case SIGQUIT:
+    if (get_wow(process, is_wow, bitness) &&
+        0 /* exit code 0 = success */ == kill_via_console_helper(bitness, L"CtrlRoutine",
+                                                                 signo == SIGINT ? CTRL_C_EVENT : CTRL_BREAK_EVENT, pid))
+      return 0;
+    /* fall-through */
+  case SIGTERM:
+    if (get_wow(process, is_wow, bitness) &&
+        0 /* exit code 0 = success */ == kill_via_console_helper(bitness, L"ExitProcess", exit_code, pid))
+      return 0;
+    break;
+  default:
+    break;
+  }
+
+  return int(TerminateProcess (process, exit_code));
+}
+
+#include <tlhelp32.h>
+#include <unistd.h>
+
+/**
+ * Terminates the process corresponding to the process ID and all of its
+ * directly and indirectly spawned subprocesses using the
+ * exit_one_process() function.
+ */
+static int
+exit_process_tree(HANDLE main_process, int exit_code)
+{
+  HANDLE snapshot = CreateToolhelp32Snapshot (TH32CS_SNAPPROCESS, 0);
+  PROCESSENTRY32 entry;
+  DWORD pids[16384];
+  int max_len = sizeof (pids) / sizeof (*pids), i, len, ret = 0;
+  pid_t pid = GetProcessId (main_process);
+  int signo = exit_code & 0x7f;
+
+  pids[0] = (DWORD) pid;
+  len = 1;
+
+  /*
+   * Even if Process32First()/Process32Next() seem to traverse the
+   * processes in topological order (i.e. parent processes before
+   * child processes), there is nothing in the Win32 API documentation
+   * suggesting that this is guaranteed.
+   *
+   * Therefore, run through them at least twice and stop when no more
+   * process IDs were added to the list.
+   */
+  for (;;)
+    {
+      memset (&entry, 0, sizeof (entry));
+      entry.dwSize = sizeof (entry);
+
+      if (!Process32First (snapshot, &entry))
+        break;
+
+      int orig_len = len;
+      do
+        {
+	  /**
+	   * Look for the parent process ID in the list of pids to kill, and if
+	   * found, add it to the list.
+	   */
+          for (i = len - 1; i >= 0; i--)
+            {
+              if (pids[i] == entry.th32ProcessID)
+                break;
+              if (pids[i] != entry.th32ParentProcessID)
+	        continue;
+
+              /* We found a process to kill; is it an MSYS2 process? */
+	      pid_t cyg_pid = cygwin_winpid_to_pid(entry.th32ProcessID);
```

Comment:
This line has funny indentation.

Reply:
I think I am going to run this file through clang-format, since this file is something we created and shouldn't create any problems with upstream.